test/k8sT/CustomCalls.go

```diff
@@ -0,0 +1,362 @@
+			serverIdentity string, direction customCallDirection) {
+
+			// Clean up tail call map entry
+			cmd := fmt.Sprintf("bpftool map delete pinned %scilium_calls_custom_%05d key %d 0 0 0", bpffsDir, endpointId, direction)
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to remove reference to byte-counter program from tail call map ('%s'): %s", cmd, res.Stderr())
+			}
+
+			// Reset byte-counter entry
+			cmd = fmt.Sprintf("bpftool map delete pinned %s key %s", mapPinPath, serverIdentity)
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to delete entry from byte-counter map ('%s'): %s", cmd, res.Stderr())
+			}
+		}
+
+		cleanupLoadedObjects := func(ciliumPod string) {
+			By("Cleaning up pinned artefacts")
+
+			// Clean up custom program
+			cmd := "rm -- " + progPinPath
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to unpin byte-counter program ('%s'): %s", cmd, res.Stderr())
+			}
+
+			// Clean up map for custom program
+			cmd = "rm -- " + mapPinPath
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to unpin byte-counter map ('%s'): %s", cmd, res.Stderr())
+			}
+			cmd = "rmdir -- " + mapsPinDir
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to remove directory for byte-counter pinned map ('%s'): %s", cmd, res.Stderr())
+			}
+		}
+
+		It("Loads byte-counter and gets consistent values", func() {
+
+			// Deploy Cilium, enable tail calls to custom programs
+			deploymentManager.DeployCilium(map[string]string{
+				"customCalls.enabled": "true",
+			}, DeployCiliumOptionsAndDNS)
+
+			// Get Cilium pod on node 2
+			ciliumPodK8s2, err := kubectl.GetCiliumPodOnNode(helpers.K8s2)
+			Expect(err).ShouldNot(HaveOccurred(), "Cannot get cilium pod on k8s2")
+
+			copyAndLoadObjectFile(ciliumPodK8s2)
+			defer cleanupLoadedObjects(ciliumPodK8s2)
+
+			getPodsInfo()
+
+			// Get ID for the endpoint for which we count the bytes
+			endpoint, err := kubectl.GetCiliumEndpoint(helpers.DefaultNamespace, podApp2.Name)
+			Expect(err).Should(BeNil(), fmt.Sprintf("Failed to retrieve endpoint for pod %s: %s", podApp2.Name, err))
+			Expect(endpoint).ShouldNot(BeNil(), fmt.Sprintf("Retrieved empty endpoint id for pod %s", podApp2.Name))
+			endpointId := endpoint.ID
+
+			// Get the identity of the pod with which the monitored
+			// pod communicates. This identity is used as a key in
+			// the byte-counter hash map.
+			identityKey := getIdentityKey("k8s:id=app1", ciliumPodK8s2)
+
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```

Comment:
There's a lot of code duplication between the two `It` blocks. I'm wondering if we could move most of the above code to a `BeforeAll` or `BeforeEach`. I'd keep the `DeployCilium` in the `It` block though.

Is it possible to install the custom program before Cilium is deployed?

Reply:
Currently the program is copied to the Cilium pod and loaded from there, so this sounds unlikely. But it may be enough to load it on the node itself, in which case this could indeed be simplified in the future.

I tried to factor as much as I could, but `copyAndLoadObjectFile()`, `getPodsInfo()`, `kubectl.GetCiliumEndpoint()`, `getIdentityKey()` all need Cilium to be deployed at the moment. One option could be to have a function that takes the options for deploying Cilium, and the two `It()` blocks simply calling it with different parameters.

Reply:
> One option could be to have a function that takes the options for deploying Cilium, and the two `It()` blocks simply calling it with different parameters.

:+1: